lib/Sema/SemaBounds.cpp

```diff
@@ -345,3 +345,86 @@ void Sema::CheckTopLevelBoundsDecls(VarDecl *D) {
   if (!D->isLocalVarDeclOrParm())
     CheckBoundsDeclarations(*this).TraverseVarDecl(D);
 }
+
+
+namespace {
+  class NonModifiyingExprSema : public RecursiveASTVisitor<NonModifiyingExprSema> {
+
+  private:
+    // Represents which kind of modifying expression we have found
+    enum ModifyingExprKind {
+      MEK_Assign,
+      MEK_Increment,
+      MEK_Decrement,
+      MEK_Call,
+      MEK_Comma,
+      MEK_Volatile
+    };
+
+  public:
+    NonModifiyingExprSema(Sema &S) : S(S) {}
+
+    // Assignments are of course modifying
+    bool VisitBinAssign(BinaryOperator* E) {
+      addError(E, MEK_Assign);
+      return false;
+    }
+
+    // Assignments are of course modifying
+    bool VisitCompoundAssignOperator(CompoundAssignOperator *E) {
+      addError(E, MEK_Assign);
+      return false;
+    }
+
+    // Pre-increment/decrement, Post-increment/decrement
+    bool VisitUnaryOperator(UnaryOperator *E) {
+      if (E->isIncrementDecrementOp()) {
+        addError(E,
+          E->isIncrementOp() ? MEK_Increment : MEK_Decrement);
+        return false;
+      }
+
+      return true;
+    }
+
+    // References to volatile variables
+    bool VisitDeclRefExpr(DeclRefExpr *E) {
+      QualType RefType = E->getType();
+      if (RefType.isVolatileQualified()) {
+        addError(E, MEK_Volatile);
+        return false;
+      }
+
+      return true;
+    }
+
+    // Function Calls are defined as modifying
+    bool VisitCallExpr(CallExpr *E) {
+      addError(E, MEK_Call);
+      return false;
+    }
+
+    // Comma Expressions are defined as modifying
+    bool VisitBinComma(BinaryOperator *E) {
+      addError(E, MEK_Comma);
```

Comment:
I believe before semantic analysis happens, that comma expressions in parameter lists are turned into a vector of arguments within a call expression. This is why calling `dynamic_check(I++, I)` will error with "too many arguments", and not with "comma expression in dynamic check expression". I can pull this out and still only prevent it at the top level.

To be honest, to even test this I had to turn off warnings about unused expressions, so maybe it's not so important, as anything truly non-modifying below this comma operator in the binary tree will be caught anyway.